marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/builtins/steps/mapping/entity-services/entity-validation-lib.sjs

```diff
@@ -0,0 +1,223 @@
+/**
+ * Main function in this library for validating an entity. If any validation errors are found and the "validateEntity"
+ * is set to "accept", then errors will be added to the options object under the "headers.datahub.validationErrors" key.
+ * This allows for them to be added to the envelope for an entity.
+ *
+ * If "validateEntity" is set to "reject", then an error will be thrown by this function.
+ *
+ * @param newInstance
+ * @param options
+ * @param entityInfo
+ */
+function validateEntity(newInstance, options, entityInfo) {
+  if (shouldValidateEntity(options)) {
+    let value = fn.string(options.validateEntity).toLowerCase();
+    if ("xml" == options.outputFormat) {
+      validateXmlEntity(newInstance, options, value);
+    } else {
+      validateJsonEntity(newInstance, options, value, entityInfo);
+    }
+  }
+}
+
+function shouldValidateEntity(options) {
+  let value = options.validateEntity;
+  if (value != null && value != undefined) {
+    value = fn.string(value).toLowerCase();
+    return value == "accept" || value == "reject";
+  }
+  return false;
+}
+
+function validateJsonEntity(newInstance, options, validateEntityValue, entityInfo) {
+  // As of 5.1.0, this is safe to do. But eventually, we'll want to find a schema by querying the schema database, or,
+  // better yet, via some API function call that does the work for us.
+  const entitySchemaUri = "/entities/" + entityInfo.title + ".entity.schema.json";
+  try {
+    xdmp.jsonValidate(newInstance, entitySchemaUri, ["full"]);
+  } catch (e) {
+    if ("accept" == validateEntityValue) {
+      if (options.headers == null) {
```

Comment:
options null check please, or default obj it

Reply:
options is guaranteed to not be null here.

Reply:
options.headers will throw a runtime error if options isn't a) set and b) an object.